features/steps/component.go

```diff
@@ -2,235 +2,102 @@ package steps
-	Version   string = ""
+	BuildTime = strconv.Itoa(time.Now().Nanosecond())
+	GitCommit = "component test commit"
+	Version   = "component test version"
 )
 
 type Component struct {
 	componenttest.ErrorFeature
-	ApiFeature        *componenttest.APIFeature
-	errorChan         chan error
-	DatasetAPI        *httpfake.HTTPFake
-	CantabularClient  *mock.CantabularClient
-	svc               *service.Service
-	cfg               *config.Config
-	wg                *sync.WaitGroup
-	signals           chan os.Signal
-	ctx               context.Context
-	HTTPServer        *http.Server
-	store             service.Datastore
-	g                 service.Generator
-	shutdownInitiated bool
-	consumer          kafka.IConsumerGroup
-	waitEventTimeout  time.Duration
+	AuthFeature       *componenttest.AuthorizationFeature
+	APIFeature        *componenttest.APIFeature
+	MongoFeature      *MongoFeature
+	DatasetFeature    *DatasetFeature
+	CantabularFeature *CantabularFeature
+
+	svc *service.Service
 }
 
-func NewComponent(t *testing.T, zebedeeURL, mongoAddr string) (*Component, error) {
-	ctx := context.Background()
+func NewComponent(t *testing.T) *Component {
+	g := &mock.Generator{URLHost: "http://mockhost:9999"}
 	cfg, err := config.Get()
 	if err != nil {
-		return nil, fmt.Errorf("failed to get config: %w", err)
+		t.Fatalf("failed to get config: %s", err)
 	}
 
-	cfg.ZebedeeURL = zebedeeURL
-	cfg.Mongo.ClusterEndpoint = mongoAddr
-	cfg.Mongo.Database = utils.RandomDatabase()
-
-	g := &mock.Generator{
-		URLHost: "http://mockhost:9999",
+	component := &Component{
+		ErrorFeature:      componenttest.ErrorFeature{TB: t},
+		AuthFeature:       componenttest.NewAuthorizationFeature(),
+		DatasetFeature:    NewDatasetFeature(t, cfg),
+		CantabularFeature: NewCantabularFeature(),
 	}
+	component.MongoFeature = NewMongoFeature(component.ErrorFeature, g, cfg)
+	component.APIFeature = componenttest.NewAPIFeature(component.Router)
 
-	mongoClient, err := GetWorkingMongo(ctx, cfg, g)
-	if err != nil {
-		return nil, fmt.Errorf("failed to create new mongo mongoClient: %w", err)
-	}
-
-	return &Component{
-		errorChan:  make(chan error),
-		wg:         &sync.WaitGroup{},
-		ctx:        ctx,
-		HTTPServer: &http.Server{},
-		cfg:        cfg,
-		DatasetAPI: httpfake.New(httpfake.WithTesting(t)),
-		CantabularClient: &mock.CantabularClient{
-			OptionsHappy: true,
-		},
-		store:            mongoClient,
-		g:                g,
-		waitEventTimeout: WaitEventTimeout,
-	}, nil
-}
-
-// Init initialises the server, the mocks and waits for the dependencies to be ready
-func (c *Component) Init() (http.Handler, error) {
-	c.signals = make(chan os.Signal, 1)
-	signal.Notify(c.signals, os.Interrupt, syscall.SIGTERM)
-
-	log.Info(c.ctx, "config used by component tests", log.Data{"cfg": c.cfg})
+	cfg.ZebedeeURL = component.AuthFeature.FakeAuthService.ResolveURL("")
 
-	c.cfg.DatasetAPIURL = c.DatasetAPI.ResolveURL("")
+	component.setInitialiserMock(g)
```

Comment:
Any reason why the generator can't be part of the component struct (and not having to pass it round like this)?

Reply:
One of the goals I had was to try to reduce/minimise the sometimes large amount of 'state' that was being stored in the component object - really all I wanted to see in the object was the service (the actual component under test) and the Features that are used as the service's dependencies (or non dependency essential features such as the API/Auth/ErrorFeature). I removed the generator because once it has been set in setInitialiserMock() is is never used again, and storing it just seemed redundant (and maybe, as in my case, can seem confusing as to why it's there)